New Excercises (#30)
Akuli/python-tutorial: basics/answers.md

        print("Access denied.")
    ```

    Again, this is not a good way to ask a real password from the user. 

> 7. Here, we create  a new variable `pallindrome_check`.We use this variable to store the  reversed  string(of`pallindrome_input`). <br>Then we check if the string that was given(`pallindrome_input`)

Reviewer: Can you delete `<br>` and fix this typo: pallindrome --> palindrome

   and the reversed sting(`pallindrome_check`) is the same using a simple `if loop`
   ```python

>    pallindrome_input=input("Type the number to check:") #to get the input from user

Reviewer: Don't write comments like this. The tutorial explains that `input(...)` gets input from the user. There is no need to explain it every time we use the function. If you want to use something that isn't explained in the tutorial, then add something to the tutorial explaining it. Most of your other comments are similar.

However, you should add comments to mention why you need to add `+2` when doing `rows+2`. That's not obvious without a comment.

   pallindrome_check=pallindrome_input[::-1] #Reverses the string
   if pallindrome_input==pallindrome_check:

>     print(f"This number is a pallindrome")

Reviewer: Please indent these with 4 spaces.

Your markdown list item (starting with `7.`) is indented with 3 spaces. That too should be indented with 4 spaces, so you get 4 spaces for most lines and 4+4 lines for indented code.

   else:

>     print("This number is not a pallindrome")

Reviewer: Did you mean "this **string** is not a palindrome"?

    print(message, "!!!")
    print(message, "!!!")
    ```

> 3. Here, we create  a new variable `palindrome_check`.We use this variable to store the  reversed  string(of`palindrome_input`). Then we check if the string that was given(`palindrome_input`)
>    and the reversed sting(`palindrome_check`) is the same using a simple `if loop`

Reviewer: I thought I added a review comment about this last time, but apparently I forgot:
- This explanation hasn't gotten outdated as the code below has changed.
- It is not an if loop.
- We should mention what `[::-1]` does, even though the reader is expected to have googled it.

```suggestion
3. In the code below, `palindrome_input[::-1]` is the string `palindrome_input` reversed.
   For example, if `palindrome_input` is `"hello"`, then `palindrome_input[::-1]` is `"olleh"`.
```

   ```python
   palindrome_input=input("Type the number to check:") 
   if palindrome_input == palindrome_input[::-1]:

>     print(f"This number is a palindrome")

Reviewer: ```suggestion
       print("This string is a palindrome")
```

   else:

>     print("This number is not a palindrome")

Reviewer: ```suggestion
       print("This string is not a palindrome")
```

    print(message, "!!!")
    print(message, "!!!")
    ```
3. Here, we create  a new variable `palindrome_check`.We use this variable to store the  reversed  string(of`palindrome_input`). Then we check if the string that was given(`palindrome_input`)
   and the reversed sting(`palindrome_check`) is the same using a simple `if loop`
   ```python
   palindrome_input=input("Type the number to check:") 

>    palindrome_check=palindrome_input[::-1] #Reverses the string

Reviewer: Please explain `[::-1]` in `handy-stuff-strings.md` or add something like "Hint: google how to reverse a string" to the exercise.

>    palindrome_check=palindrome_input[::-1] #Reverses the string
>    if palindrome_input==palindrome_check:

Reviewer: We don't need an unnecessary variable. We just want to check if the string itself is same as after reversing.

```suggestion
   if palindrome_input == palindrome_input[::-1]:
```

The explanation should be updated too.

    print(converted_numbers)
    ```

5. ``` python
   row_count = int(input("Type the number of rows needed:")) 
   for column_count in range(1, row_count+1):
       # Print numbers from 1 to column_count
       for number in range(1, column_count+1): 
           print(number, end=" ") 
       print()  # creates a new line for the next row
   ```

>    We actually use the first `for` loop to loop through the number of rows needed.Now,
>    through that `for` loop we got to which column the number should be inserted.
>    Now,the next `for` loop helps us to identify which numbers should be printed on the row which we currently accessed from the `for` loop above

Reviewer: This explanation hasn't changed much during the reviewing, so I just rewrote it, lol. What do you think?

```suggestion
   If the user enters 5, we want to do a row with 1 column, then 2 columns, and so on until 5 columns.
   That would be `for column_count in range(1, 6)`, because the end of the range is excluded.
   In general, we need to specify `row_count + 1` so that it actually ends at  `row_count`.
   The second loop is similar.

   Usually `print(number)` puts a newline character at the end of the line, so that the next print goes to the next line.
   To get all numbers on the same line, we use a space instead of a newline character,
   but we still need `print()` to add a newline character once we have printed the entire row.
```

Author: Thank You

Author: It is more than fine

   
  
6. ```python
   row_count=int(input("Type the number of rows needed:"))
    

>    for column_count in range(1, row_count + 1): # first for loop
>      for number in range(column_count,row_count+1):# second for loop 

Reviewer: I don't think `column_count` is a good variable name. If it is 1, we print 1,2,3,4,5, so 5 columns. Not 1 column, like you would expect.

I also changed the text below so it doesn't refer to "first for loop" etc, so the comments are no longer needed.

```suggestion
   for line_number in range(1, row_count + 1):
     for number in range(line_number, row_count+1):
```

        print(number, end=' ')
     print()
   ```

>    Here,the first `for` loop fetches us the column to which we have to add the number.<br>

Reviewer: ```suggestion
   Just like in the previous exercise, if the user enters 5, the first `for` loop gives the line numbers `1, 2, 3, 4, 5`.<br>
```

>    Now,suppose the column is `2`.The second `for` loop finds all the numbers that lie between the column number(i.e. here 2) and the total rows we need.
>    Eg:If total rows we need is 5 then we get 2,3,4 and 5 in the second `for` loop, when column_count is 2<br>

Reviewer: ```suggestion
   For example, on line 2, we should print numbers from 2 to 5, as in `range(2, 6)`, or in general, `range(line_number, row_count+1)`.
```

    print(converted_numbers)
    ```

5. ``` python
   row_count = int(input("Type the number of rows needed:")) 
   for column_count in range(1, row_count+1):

>        for number in range(1, column_count+1): 

Reviewer: ```suggestion
       # Print numbers from 1 to column_count
       for number in range(1, column_count+1): 
```

           print(number, end=" ") 
       print()  # creates a new line for the next row
   ```
   We actually use the first `for` loop to loop through the number of rows needed.Now,
   through that `for` loop we got to which column the number should be inserted.
   Now,the next `for` loop helps us to identify which numbers should be printed on the row which we currently accessed from the `for` loop above

   
  
6. The idea is more or less same as in the above question.
   ```python
   row_count=int(input("Type the number of rows needed:"))
    

>    for column_count in range(1, row_count + 1): 
>      for number in range(row_count,column_count+1): 

Reviewer: This code still doesn't make sense to me when I read it. It also doesn't work.

The range must start from row number, which goes from 1 to row count. So how about this?

```suggestion
   for row_number in range(1, row_count + 1):
     for number in range(row_number, row_count+1):
```

It still needs a better explanation. "More or less same as above" is not enough, because we now have to start the range from `row_number`, not from `1`. For example, on row number 2, we want to print numbers from 2 to 5. It is what the range does, but for someone who is new to ranges, that isn't obvious.

        converted_numbers.append(int(number))
    print(converted_numbers)
    ```

> 5. We actually use the first `for` loop to loop through the number of rows needed.Now,

Reviewer: This line seems to refer to the first line of the code below it. If you want to write the explanation like that, then show it first, so that it can be read easily from top to bottom.

   through that `for` loop we got the row number.<br>
   Now,the next `for` loop helps us to identify which numbers should be printed on the row which we currently accessed from the `for` loop above
   ```python
   rows=int(input("Type the number of rows needed:")) #gets the number of rows from the user
   
   for row in range(1,rows+1): #finds the row from rows' variable
    for column in range(1,row+1): # this is actually the column where a number has to be added

>         print(column,end=" ") # this is where the number is printed into rows

Reviewer: I have not shown `end=" "` in this tutorial. Can you add some text before the exercises section that explains it?

Author: Ok

Author: ![image](https://user-images.githubusercontent.com/75750164/103444835-c80a8180-4c92-11eb-9aca-ec2509909f81.png)
I can add this in this section right?
This is [using-functions.md](https://github.com/Akuli/python-tutorial/blob/master/basics/using-functions.md)

Reviewer: That seems like a good place.

    print(converted_numbers)
    ```

5. ``` python

>    rows=int(input("Type the number of rows needed:")) 
>    for row in range(1,rows+1):
>        for column in range(1,row+1): 
>            print(column,end=" ") 
>        print()#creates a new line for the next row

Reviewer: There's an explanation below where you basically explain what `rows` and `row` mean. I think it would be better to use better variable names instead. (I also added some whitespace.)

```suggestion
   row_count = int(input("Type the number of rows needed:")) 
   for column_count in range(1, row_count+1):
       for number in range(1, column_count+1): 
           print(number, end=" ") 
       print()  # creates a new line for the next row
```

You will also need to change the other answer and update the explanations.
